generators/chipyard/src/main/scala/harness/HasHarnessInstantiators.scala

```diff
@@ -0,0 +1,96 @@
+import scala.collection.mutable.{ArrayBuffer, LinkedHashMap}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import org.chipsalliance.cde.config.{Field, Parameters, Config}
+import freechips.rocketchip.util.{ResetCatchAndSync}
+import freechips.rocketchip.prci.{ClockBundle, ClockBundleParameters, ClockSinkParameters, ClockParameters}
+import freechips.rocketchip.stage.phases.TargetDirKey
+
+import chipyard.harness.{ApplyHarnessBinders, HarnessBinders}
+import chipyard.iobinders.HasIOBinders
+import chipyard.clocking.{SimplePllConfiguration, ClockDividerN}
+import chipyard.{ChipTop}
+
+// -------------------------------
+// Chipyard Test Harness
+// -------------------------------
+
+case object MultiChipNChips extends Field[Int](0) // 0 means ignore MultiChipParams
+case class MultiChipParameters(chipId: Int) extends Field[Parameters]
+case object BuildTop extends Field[Parameters => LazyModule]((p: Parameters) => new ChipTop()(p))
+case object HarnessClockInstantiatorKey extends Field[() => HarnessClockInstantiator]()
+case object HarnessBinderClockFrequencyKey extends Field[Double](100.0) // MHz
+case object MultiChipIdx extends Field[Int](0)
+
+class WithMultiChip(id: Int, p: Parameters) extends Config((site, here, up) => {
+  case MultiChipParameters(`id`) => p
+  case MultiChipNChips => up(MultiChipNChips) max (id + 1)
+})
+
+class WithHomogeneousMultiChip(n: Int, p: Parameters, idStart: Int = 0) extends Config((site, here, up) => {
+  case MultiChipParameters(id) => if (id >= idStart && id < idStart + n) p else up(MultiChipParameters(id))
+  case MultiChipNChips => up(MultiChipNChips) max (idStart + n)
+})
+
+class WithHarnessBinderClockFreqMHz(freqMHz: Double) extends Config((site, here, up) => {
+  case HarnessBinderClockFrequencyKey => freqMHz
+})
+
+// A TestHarness mixing this in will
+// - use the HarnessClockInstantiator clock provide
+trait HasHarnessInstantiators {
+  implicit val p: Parameters
+  // clock/reset of the chiptop reference clock (can be different than the implicit harness clock/reset)
+  private val harnessBinderClockFreq: Double = p(HarnessBinderClockFrequencyKey)
+  def getHarnessBinderClockFreqHz: Double = harnessBinderClockFreq * 1000000
+  def getHarnessBinderClockFreqMHz: Double = harnessBinderClockFreq
+
+  // buildtopClock takes the refClockFreq, and drives the harnessbinders
+  val harnessBinderClock = Wire(Clock())
+  val harnessBinderReset = Wire(Reset())
+
+  // classes which inherit this trait should provide the below definitions
+  def referenceClockFreqMHz: Double
+  def referenceClock: Clock
+  def referenceReset: Reset
+  def success: Bool
+
+  // This can be accessed to get new clocks from the harness
+  val harnessClockInstantiator = p(HarnessClockInstantiatorKey)()
+
+  private val chipParameters = if (p(MultiChipNChips) == 0) {
+    Seq(p)
+  } else {
+    (0 until p(MultiChipNChips)).map { i => p(MultiChipParameters(i)).alterPartial {
+      case TargetDirKey => p(TargetDirKey) // hacky fix
```

Comment:
Why is this getting dropped? I assume the `TargetDirKey` is set in the parameters are passed in / duplicated.

Reply:
The generate phase adds the `TargetDirKey` to the `Parameters` object before it is passed into this generator., but not in the nested `Parameters` within `MultiChipParameters`. 

This propagates `TargetDirKey` down to the nested parametesr in `MultiChipParameters`.